### PR TITLE
Initial slacklib

### PR DESCRIPTION
**Pull request by jupierce**
Implements a few APIs to send slack messages. The most common pattern should be:
```
slack = slacklib.to('4.5')  // direct output to art-release-4-5
slack.task('what you are doing') {
    ..do your work.. e.g. run doozer
    it.say('progress you want to notify people about')
}
```
- Each release has a private slack channel (`art-release-3-11`, `art-release-4-1`, etc) 
- All slack output using this pattern will go to a slack thread except for failures (i.e. an exception thrown within the closure).
- A failure will add to the main thread, but will also re-broadcast to the channel.
- Each task thread will report the Jenkins job that created it and provide a link back to the job.
- If containing Jenkins job is a manual run, the user who ran the job will receive a callout in the task's slack thread. 

Less common patterns:
```
slack = slacklib.to('#art-team')
slack_thread = slack.say('hello!`)
slack_thread.say('this will be in a thread below hello!')
try {
    ...do your work...
} catch (e) {
    slack_thread.failure('Report a failure manually')
}

```

### CODE REVIEW

File: jobs/build/release/Jenkinsfile [hunk 1]
```diff
@@ -97,7 +98,8 @@ node {
 
     buildlib.cleanWorkdir("${env.WORKSPACE}")
 
-    try {
+    jobThread = slacklib.to(FROM_RELEASE_TAG)
+    jobThread.task("Public release prep for: ${FROM_RELEASE_TAG}") {
```
Comment (sosiouxme) [inline jobs/build/release/Jenkinsfile]: replaces try/catch/email... :heart:

File: pipeline-scripts/slacklib.groovy [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+            try {
+                def parts = channel_or_release.split('\\D')  // split on any non-digit; '4.2.5' => ['4','2','5']
+                if ( parts[0] == '3' || parts[0] == '4' ) {
+                    // Turn into a release channel name: e.g. "4.1" -> "#art-release-4-1"
+                    channel = "#art-release-${parts[0]}-${parts[1]}"
+                } else {
+                    echo "ERROR Unexpected fields in ${channel_or_release}"
+                }
+            } catch ( pe ) {
+                echo "ERROR determining channel for ${channel_or_release}; using default ${getChannelDefault()}: ${e}"
```
Comment (sosiouxme) [inline pipeline-scripts/slacklib.groovy]:
```suggestion
                echo "ERROR determining channel for ${channel_or_release}; using default ${getChannelDefault()}: ${pe}"
```

File: pipeline-scripts/slacklib.groovy [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+def to(channel_or_release, verbose=false, as_user='art-release-bot') {
+    def channel = channel_or_release
+    if ( channel_or_release.charAt(0) != '#' ) {
+        channel = getDefaultChannel()
+        if ( channel_or_release.charAt(0).isDigit() ) { // something like 4.5?
+            try {
+                def parts = channel_or_release.split('\\D')  // split on any non-digit; '4.2.5' => ['4','2','5']
+                if ( parts[0] == '3' || parts[0] == '4' ) {
+                    // Turn into a release channel name: e.g. "4.1" -> "#art-release-4-1"
+                    channel = "#art-release-${parts[0]}-${parts[1]}"
```
Comment (sosiouxme) [inline pipeline-scripts/slacklib.groovy]: do we need to pre-create these or do they spring into existence?

**Comment by sosiouxme** [comment]
also if running under hack namespace, would want to automatically note that... and link the specific job...

File: pipeline-scripts/slacklib.groovy (hunk 3 above)
Comment (jupierce) [reply pipeline-scripts/slacklib.groovy]: I've already created them. Will invite if we merge. Didn't want to jump the gun if the PR was not what you wanted.

**Comment by jupierce** [comment]
It has been run under hack: https://saml.buildvm.openshift.eng.bos.redhat.com:8888/job/hack/job/jupierce2/64/console . Replay the last job to see the code. The job looks like it failed, but that is because the code was testing throwing errors in `task(){}`.

File: pipeline-scripts/slacklib.groovy (hunk 2 above)
Comment (jupierce) [reply pipeline-scripts/slacklib.groovy]: Nice catch!

**Comment by sosiouxme** [comment]
now that i can see example output, ready to merge :)
i see the job is linked, which is the main thing. i was thinking it should also say `[hack]` or something next to it so we could visually identify that this was a run from a hack job rather than the real thing. that can be a later refinement though.

**Pull request merged**